github/codeowners.py

```diff
@@ -116,38 +118,146 @@ def resolve_team_members(
-            yield member.email
+    yield from map(github_api.user, team_or_none.members())
+
+
+@dataclasses.dataclass
+class CodeownerEntry:
+    value: str
+
+
+@dataclasses.dataclass
+class UsernameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class EmailCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class TeamnameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+class Filter:
+    @staticmethod
+    def usernames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, UsernameCodeownerEntry)
+
+    @staticmethod
+    def emails(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, EmailCodeownerEntry)
+
+    @staticmethod
+    def teamnames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, TeamnameCodeownerEntry)
+
+
+def iter_codeowners(
+    codeowners_entries,
+    node_filter: typing.Callable[[CodeownerEntry], bool]=None,
+) -> typing.Generator[str, None, None]:
+    '''
+    returns a generator yielding the codeowner entries.
+
+    @param codeowner_entries:  root for iteration
+    @node_filter:              use to filter emitted nodes (see Filter for predefined filters)
+    '''
+
+    for codeowners_entry in codeowners_entries:
+        if '@' not in codeowners_entry:
+            logger.warning(f'invalid codeowners-entry: {codeowners_entry}')
+            continue
+
+        if not codeowners_entry.startswith('@'):
+            node = EmailCodeownerEntry(value=codeowners_entry)
+
+        elif '/' not in codeowners_entry:
+            node = UsernameCodeownerEntry(value=codeowners_entry[1:])
+
         else:
-            logger.warning(f'no email found for GitHub user {member}')
+            node = TeamnameCodeownerEntry(value=codeowners_entry[1:])
+
+        if node_filter and not node_filter(node):
+            continue
+
+        yield node.value
 
 
 def resolve_email_addresses(
-    codeowners_entries,
+    codeowners_entries: typing.Iterable[str],
     github_api: GitHub,
 ) -> typing.Generator[str, None, None]:
     '''
-    returns a generator yielding the resolved email addresses for the given iterable of
-    github codeowners entries.
+    returns a generator yielding unique email addresses resolved from codeowner entries.
     '''
-    for codeowner_entry in codeowners_entries:
-        if '@' not in codeowner_entry:
-            logger.warning(f'invalid codeowners-entry: {codeowner_entry}')
-            continue
-        if not codeowner_entry.startswith('@'):
-            yield codeowner_entry # plain email address
-        elif '/' not in codeowner_entry:
-            email_addr = determine_email_address(
-                github_user_name=codeowner_entry[1:],
-                github_api=github_api,
-            )
-            if email_addr:
-                yield email_addr
+
+    # iter multiple times
+    codeowners_entries = list(codeowners_entries)
```

Comment:
if mutability is not required, prefer tuple
why do we need to iterate multiple times?